[DTensor][Partial] fixing adding scalar to Partial

anshul-si opened the pull request:
**Summary:** This is a fix for adding a scalar to a Partial dTensor reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This is accomplished when checking if we should add a Partial placement. If one of the arguments is a scalar and the op is aten.add_Tensor, we should redistribute the Partial to replicate. 

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_partial_scalar

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #167813



cc @wanchaol @tianyu-l @wz337 @XilunWu @d4l3k @pragupta @SherlockNoMad @H-Huang @awgu @fegin @fduwjj @wconstab @msaroufim @dcci

anshul-si committed:
[DTensor][Partial] fixing adding scalar to Partial

[ghstack-poisoned]

anshul-si committed:
Update on "[DTensor][Partial] fixing adding scalar to Partial"

**Summary:** This is a fix for adding a scalar to a Partial dTensor reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. We accomplish this by adding a function that checks if we support Partial placements when attempting to add them to the output_spec. Regarding the specific checks for adding a scalar to Partial, I check if aten.add.Tensor is the op, and if so, is there a scalar argument in arg_schemas. If this is the case, I return False to force replication. Currently, we don't need to do this for aten.mul.Tensor as it works for all the reduction ops for each Partial placement. However, in the future, new Partial placements could be added where the reduce op requires redistribution. In this case, I replicate, but I warn the user I'm replicating.

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_partial_scalar
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_unverified_custom_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad H-Huang awgu fegin fduwjj wconstab msaroufim dcci

[ghstack-poisoned]

wconstab commented on torch/distributed/tensor/_ops/_pointwise_ops.py:
```diff
@@ -488,7 +493,54 @@ def linear_pointwise_strategy(op_schema: OpSchema) -> StrategyType:
+        op in (addition_ops + multiplication_ops)
+        and type(placement) not in verified_partial_placements
+    ):
+        warnings.warn(
```
we can't do this on every op. it should at most be a 'warn_once' call

wconstab reviewed (commented):
my suggestion would be to simplify this PR to just the minimum fix for the scalar+partial bugs, to limit the chance of coupling the bugfix to an unintended side effect.  if you have a separate bug you're trying to fix or a refactor, make it a separate PR and we can discuss it on its own merits

anshul-si committed:
Update on "[DTensor][Partial] fixing adding scalar to Partial"

**Summary:** This is a fix for adding a scalar to a Partial dTensor reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. We accomplish this by adding a function that checks if we support Partial placements when attempting to add them to the output_spec. Regarding the specific checks for adding a scalar to Partial, I check if aten.add.Tensor is the op, and if so, is there a scalar argument in arg_schemas. If this is the case, I return False to force replication. Currently, we don't need to do this for aten.mul.Tensor as it works for all the reduction ops for each Partial placement. However, in the future, new Partial placements could be added where the reduce op requires redistribution. In this case, I replicate, but I warn the user I'm replicating.

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_partial_scalar
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_unverified_custom_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad H-Huang awgu fegin fduwjj wconstab msaroufim dcci

[ghstack-poisoned]

anshul-si committed:
Update on "[DTensor][Partial] fixing adding scalar to Partial"

**Summary:** This is a fix for adding a scalar to a Partial dTensor reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. We accomplish this by adding a function that checks if we support Partial placements when attempting to add them to the output_spec. Regarding the specific checks for adding a scalar to Partial, I check if aten.add.Tensor is the op, and if so, is there a scalar argument in arg_schemas. If this is the case, I return False to force replication. Currently, we don't need to do this for aten.mul.Tensor as it works for all the reduction ops for each Partial placement. However, in the future, new Partial placements could be added where the reduce op requires redistribution. In this case, I replicate, but I warn the user I'm replicating.

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_partial_scalar
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_unverified_custom_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad H-Huang awgu fegin fduwjj wconstab msaroufim dcci

[ghstack-poisoned]

anshul-si committed:
Update on "[DTensor][Partial] fixing adding scalar to Partial"

**Summary:** This is a fix for adding a scalar to a Partial dTensor reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. We accomplish this by adding a function that checks if we support Partial placements when attempting to add them to the output_spec. Regarding the specific checks for adding a scalar to Partial, I check if aten.add.Tensor is the op, and if so, is there a scalar argument in arg_schemas. If this is the case, I return False to force replication. Currently, we don't need to do this for aten.mul.Tensor as it works for all the reduction ops for each Partial placement. However, in the future, new Partial placements could be added where the reduce op requires redistribution. In this case, I replicate, but I warn the user I'm replicating.

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_partial_scalar
2. pytest test/distributed/tensor/test_pointwise_ops.py -k test_unverified_custom_partial




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad H-Huang awgu fegin fduwjj wconstab msaroufim dcci

[ghstack-poisoned]

wconstab commented on torch/distributed/tensor/_ops/_pointwise_ops.py:
```diff
@@ -530,10 +533,16 @@ def common_pointwise_strategy(
                 out_placements.append(Shard(new_shard_dim))
             elif isinstance(placement, Partial):
+                # partial + scalar doesn't work
+                addition_ops = [aten.add.Tensor, aten.add_.Tensor]
```
i think i would feel better about having the logic be 
* by default, Op(partial, scalar) is not supported (forces redistribute)
* for known good combinations of Partial type and operator, we opt into allowing partial,scalar -> partial sharding rule

its less likely we have a hole in our logic that way.  the way it is now, we're only plugging the holes we thought of, and i'm not sure we got them all.

anshul-si replied:
For the existing linear ops:

aten.to.dtype 
composes to aten._to_copy.default, we should never have to worry about this case. However aten._to_copy.default with Partial itself is broken. I will address this in another PR.
<img width="4032" height="1067" alt="image" src="https://github.com/user-attachments/assets/b1cfa7d7-9d91-4ba5-bc3c-29a08da8dc37" />

aten.div.Scalar
aten.div_.Scalar
aten.mul.Scalar
aten.mul_.Scalar 
Their linearity set to 0. When checking to add Partial placement, if linearity <= 0, we currently convert them to a Replicate. I believe this is a bug since these operations should be linear as I will prove below. However, if I'm mistaken, we should move these ops to pointwise_ops as linearity 0 has the same effect as linearity -1 currently. 
<img width="4032" height="692" alt="image" src="https://github.com/user-attachments/assets/e8a82825-5c5f-43f3-9767-43c3ece6b88f" />

aten.mul.Tensor
aten.mul_.Tensor
<img width="4032" height="1196" alt="image" src="https://github.com/user-attachments/assets/17ca7e57-93b9-4e43-9087-67f293823b40" />

aten.copy_.default
Trying to figure out how I mathematically prove this, but I haven't thought of any examples that break

anshul-si replied:
I agree with what you're saying. Assuming what i'm saying about mult / div linearity, i think those are safe to make exceptions for.

anshul-si committed:
Update on "[DTensor][Partial] fixing adding scalar to Partial"

**Summary:** This is a fix for adding a scalar to a Partial dTensor reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This is accomplished when checking if we should add a Partial placement. If one of the arguments is a scalar and the op is aten.add_Tensor, we should redistribute the Partial to replicate. 

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_partial_scalar




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad H-Huang awgu fegin fduwjj wconstab msaroufim dcci

[ghstack-poisoned]

tianyu-l commented on torch/distributed/tensor/_ops/_pointwise_ops.py:
```diff
@@ -530,11 +533,25 @@ def common_pointwise_strategy(
+                partial_supports_linearity = (
+                    placement.is_partial("sum") or placement.is_partial("avg")
+                ) and (
+                    op in safe_partial_ops
```
it might be cleaner to create another linearity type, instead of requiring `op` to be passed in

ezyang commented on torch/distributed/tensor/_ops/_pointwise_ops.py:
```diff
@@ -530,11 +533,25 @@ def common_pointwise_strategy(
+                    placement.is_partial("sum") or placement.is_partial("avg")
+                ) and (
+                    op in safe_partial_ops
+                    or not any(isinstance(arg, _Number) for arg in args_schema)
```
@tianyu-l this is oooold. The add.Scalar overload was added later. Before it existed, we use this hack called "wrapped number" to indicate a Tensor came from a Python number, so it gets different semantics. The eager frontend still goes to this overload.

anshul-si commented on torch/distributed/tensor/_ops/_pointwise_ops.py:
```diff
@@ -530,11 +533,25 @@ def common_pointwise_strategy(
-                ) or placement.is_partial("avg")
-                if linearity > 0 and partial_supports_linearity:
+                partial_supports_linearity = (
+                    placement.is_partial("sum") or placement.is_partial("avg")
```
You're correct. aten.mult.scalar for NormPartial won't work the way its intended when we multiply with a negative scalar.

tianyu-l commented:
any progress on this?

anshul-si committed:
Update on "[DTensor][Partial] fixing adding scalar to Partial"

**Summary:** This is a fix for adding a scalar to a Partial dTensor reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This is accomplished when checking if we should add a Partial placement. If one of the arguments is a scalar and the op is aten.add_Tensor, we should redistribute the Partial to replicate. 

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_partial_scalar




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad H-Huang awgu fegin fduwjj wconstab msaroufim dcci

[ghstack-poisoned]

anshul-si committed:
Update on "[DTensor][Partial] fixing adding scalar to Partial"

**Summary:** This is a fix for adding a scalar to a Partial dTensor reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This is accomplished when checking if we should add a Partial placement. If one of the arguments is a scalar and the op is aten.add_Tensor, we should redistribute the Partial to replicate. 

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_partial_scalar




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad H-Huang awgu fegin fduwjj wconstab msaroufim dcci

[ghstack-poisoned]

anshul-si committed:
Update on "[DTensor][Partial] fixing adding scalar to Partial"

**Summary:** This is a fix for adding a scalar to a Partial dTensor reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This is accomplished when checking if we should add a Partial placement. If one of the arguments is a scalar and the op is aten.add_Tensor, we should redistribute the Partial to replicate. 

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_partial_scalar




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad H-Huang awgu fegin fduwjj wconstab msaroufim dcci

[ghstack-poisoned]

anshul-si committed:
Update on "[DTensor][Partial] fixing adding scalar to Partial"

**Summary:** This is a fix for adding a scalar to a Partial dTensor reported in https://github.com/pytorch/pytorch/issues/149768, https://github.com/pytorch/pytorch/issues/163193. This is accomplished when checking if we should add a Partial placement. If one of the arguments is a scalar and the op is aten.add_Tensor, we should redistribute the Partial to replicate. 

**Test Cases**
1. pytest test/distributed/tensor/test_pointwise_ops.py -k test_add_partial_scalar




cc wanchaol tianyu-l wz337 XilunWu d4l3k pragupta SherlockNoMad H-Huang awgu fegin fduwjj wconstab msaroufim dcci

[ghstack-poisoned]

anshul-si commented:
Messed up git control with ghstack, will link to new PRs.
